File: test/csrc/com/xuggle/xuggler/ContainerFormatTest.cpp
/*******************************************************************************
 * Copyright (c) 2008, 2010 Xuggle Inc.  All rights reserved.
 *  
 * This file is part of Xuggle-Xuggler-Main.
 *
 * Xuggle-Xuggler-Main is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Xuggle-Xuggler-Main is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Xuggle-Xuggler-Main.  If not, see <http://www.gnu.org/licenses/>.
 *******************************************************************************/

#include <string>
#include <com/xuggle/ferry/RefPointer.h>
#include <com/xuggle/ferry/Logger.h>
#include <com/xuggle/xuggler/IContainerFormat.h>
#include <com/xuggle/xuggler/Global.h>
#include "Helper.h"
#include "ContainerFormatTest.h"

using namespace VS_CPP_NAMESPACE;

VS_LOG_SETUP(VS_CPP_PACKAGE);

ContainerFormatTest :: ContainerFormatTest()
{
  h = 0;
}

ContainerFormatTest :: ~ContainerFormatTest()
{
  tearDown();
}

void
ContainerFormatTest :: setUp()
{
  format = 0;
  h = new Helper();
}

void
ContainerFormatTest :: tearDown()
{
  if (h)
    delete h;
  h = 0;
  format = 0;
}

void
ContainerFormatTest :: testCreationAndDestruction()
{
  format = IContainerFormat::make();
  VS_TUT_ENSURE("could not allocate container", format);
  VS_TUT_ENSURE("no one else owns this container", format->getCurrentRefCount() == 1);
}

void
ContainerFormatTest :: testSetInputFormat()
{
  int retval = -1;
  format = IContainerFormat::make();
  VS_TUT_ENSURE("could not allocate container", format);

  // Valid formats
  retval = format->setInputFormat("flv");
  VS_TUT_ENSURE("could not set format to flv", retval >= 0);
  
  // Invalid formats
  retval = format->setInputFormat("notavalidshortname");
  VS_TUT_ENSURE("could set to an invalid format", retval < 0);
  retval = format->setInputFormat("");
  VS_TUT_ENSURE("could set to an invalid format", retval < 0);
  retval = format->setInputFormat(0);
  VS_TUT_ENSURE("could set to an invalid format", retval < 0);
}

void
ContainerFormatTest :: testSetOutputFormat()
{
  int retval = -1;
  format = IContainerFormat::make();
  VS_TUT_ENSURE("could not allocate container", format);
  
  // Valid formats
  retval = format->setOutputFormat("flv", 0, 0);
  VS_TUT_ENSURE("could not set format to flv", retval >= 0);
  retval = format->setOutputFormat("flv","", "");
  VS_TUT_ENSURE("could not set format to flv", retval >= 0);    
  retval = format->setOutputFormat("flv",h->SAMPLE_FILE, "");
  VS_TUT_ENSURE("could not set format to flv", retval >= 0);
  retval = format->setOutputFormat(0,h->SAMPLE_FILE, "");
  VS_TUT_ENSURE("could not set format to flv", retval >= 0);
  retval = format->setOutputFormat("",h->SAMPLE_FILE, "");
  VS_TUT_ENSURE("could not set format to flv", retval >= 0);

  // invalid formats
  retval = format->setOutputFormat(0, 0, 0);
  VS_TUT_ENSURE("could set to an invalid format", retval < 0);
  retval = format->setOutputFormat("", "", "");
  VS_TUT_ENSURE("could set to an invalid format", retval < 0);

  // FFMPEG returns an internal format if it doesn't know
  // the short name.
  retval = format->setOutputFormat("notavalidshortname", "", "");
  VS_TUT_ENSURE("could set to an invalid format", retval == 0);
  VS_TUT_ENSURE_EQUALS("unexpected format short name",
      std::string(format->getOutputFormatShortName()),
      std::string("ffm"));
  VS_TUT_ENSURE_EQUALS("unexpected format long name",
      std::string(format->getOutputFormatLongName()),
      std::string("FFM (FFserver live feed) format"));
  VS_TUT_ENSURE("no mime type expected", format->getOutputFormatMimeType());
  
  // FFMPEG retuns an internal format if it doesn't know the
  // long name.
  retval = format->setOutputFormat("", "notavalidurl&kdjs2kd:dkjs", "");
  VS_TUT_ENSURE("could set to an invalid format", retval == 0);
  VS_TUT_ENSURE_EQUALS("unexpected format short name",
      std::string(format->getOutputFormatShortName()),
      std::string("ffm"));
  VS_TUT_ENSURE_EQUALS("unexpected format long name",
      std::string(format->getOutputFormatLongName()),
      std::string("FFM (FFserver live feed) format"));
  VS_TUT_ENSURE("no mime type expected", format->getOutputFormatMimeType());

  // But it doesn't return anything if it doesn't know the
  // mime type.
  retval = format->setOutputFormat("", "", "notavalidmeimetype/foo");
  VS_TUT_ENSURE("could set to an invalid format", retval < 0 );
}

void 
ContainerFormatTest :: testGetOutputNumCodecsSupported()
{
  LoggerStack stack;
  stack.setGlobalLevel(Logger::LEVEL_INFO, false);

  int retval = -1;
  format = IContainerFormat::make();
  VS_TUT_ENSURE("could not allocate container", format);
  
  // Valid formats
  retval = format->setOutputFormat("mov", 0, 0);
  VS_TUT_ENSURE("could not set format to mov", retval >= 0);
  
  VS_TUT_ENSURE("no default audio codec",
      format->getOutputDefaultAudioCodec() != ICodec::CODEC_ID_NONE);
  VS_TUT_ENSURE("no default audio codec",
      format->getOutputDefaultVideoCodec() != ICodec::CODEC_ID_NONE);
  VS_TUT_ENSURE("should have non null extensions",
      format->getOutputExtensions());
  int numCodecsSupported = format->getOutputNumCodecsSupported();
  VS_LOG_DEBUG("Num codecs supported: %ld", numCodecsSupported);
  VS_TUT_ENSURE("should have more than 1 codec supported",
      numCodecsSupported > 1);
  for(int i = 0; i < numCodecsSupported; i++)
  {
    ICodec::ID id = format->getOutputCodecID(i);
    VS_TUT_ENSURE("should not be none", id != ICodec::CODEC_ID_NONE);
    RefPointer<ICodec> codec = ICodec::findEncodingCodec(id);
    if (!codec) {
      VS_LOG_DEBUG("Codec not supported: %ld", id);
    } else {
      int tag = format->getOutputCodecTag(i);
      VS_LOG_DEBUG("Codec: %s; Tag: %ld; ID: %ld",
          codec->getName(),
          tag,
          id);
      VS_TUT_ENSURE("tag should be non zero", tag != 0);
      VS_TUT_ENSURE("should say we support this",
          format->isCodecSupportedForOutput(id));
      VS_TUT_ENSURE_EQUALS("tags should be the same",
          tag,
          format->getOutputCodecTag(i));
    }
  }
}

void
ContainerFormatTest :: testGetInstalledInputFormats()
{
  LoggerStack stack;
  stack.setGlobalLevel(Logger::LEVEL_INFO, false);
  int32_t count = IContainerFormat::getNumInstalledInputFormats();
  VS_TUT_ENSURE("should be more than one", count > 1);
  
  for(int i=0; i < count; i++)
  {
    RefPointer<IContainerFormat> fmt =
      IContainerFormat::getInstalledInputFormat(i);
    VS_TUT_ENSURE("should be valid", fmt);
    if (fmt) {
      VS_LOG_DEBUG("%s; %s",
          fmt->getInputFormatShortName(),
          fmt->getInputFormatLongName());
      VS_TUT_ENSURE("is input", fmt->isInput());
    }
  }
}

void
ContainerFormatTest :: testGetInstalledOutputFormats()
{
  LoggerStack stack;
  stack.setGlobalLevel(Logger::LEVEL_INFO, false);
  int32_t count = IContainerFormat::getNumInstalledOutputFormats();
  VS_TUT_ENSURE("should be more than one", count > 1);
  
  for(int i=0; i < count; i++)
  {
    RefPointer<IContainerFormat> fmt =
      IContainerFormat::getInstalledOutputFormat(i);
    VS_TUT_ENSURE("should be valid", fmt);
    if (fmt) {
      VS_LOG_DEBUG("%s; %s",
          fmt->getOutputFormatShortName(),
          fmt->getOutputFormatLongName());
      VS_TUT_ENSURE("is output", fmt->isOutput());
    }
  }
}

File: test/src/com/xuggle/xuggler/ContainerFormatTest.java
/*
 * Copyright (c) 2008, 2009 by Xuggle Incorporated.  All rights reserved.
 * 
 * This file is part of Xuggler.
 * 
 * You can redistribute Xuggler and/or modify it under the terms of the GNU
 * Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any
 * later version.
 * 
 * Xuggler is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public
 * License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with Xuggler.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.xuggle.xuggler;

import java.util.Collection;
import java.util.List;

import org.junit.*;

import com.xuggle.ferry.JNIMemoryManager;
import com.xuggle.xuggler.IContainerFormat;

import junit.framework.TestCase;

public class ContainerFormatTest extends TestCase
{
  private final String sampleFile = "fixtures/testfile.flv";

  @Test
  public void testSetOutputFmt()
  {
    IContainerFormat fmt = IContainerFormat.make();
    
    fmt.setOutputFormat("flv", null, null);
    fmt.setOutputFormat("flv", sampleFile, null);
    fmt.setOutputFormat("flv", "file:"+sampleFile, null);

    
    fmt.setOutputFormat("NotAShortName", null, null);
    fmt.setOutputFormat("NotAShortName", "NotAURL", null);
    fmt.setOutputFormat("NotAShortName", "file:"+"NotAURL", null);
    fmt.setOutputFormat("NotAShortName", "NotAProtocol:"+"NotAURL", null);

    
    assertTrue("got to end of test without coredump.  woo hoo", true);
  }
  
  @Test
  public void testSetInputFmt()
  {
    IContainerFormat fmt = IContainerFormat.make();
    
    fmt.setInputFormat("flv");
    fmt.setInputFormat("mov");
    fmt.setInputFormat("NotAShortName");
    assertTrue("got to end of test without coredump.  woo hoo", true);
  }
  
  @Test
  public void testGetInputFlag()
  {
    IContainerFormat fmt = IContainerFormat.make();
    int retval = -1;
    int flags = fmt.getInputFlags();
    assertEquals("should be not set", flags, 0);
    retval = fmt.setInputFormat("s16be");
    assertTrue("should succeed", retval >= 0);
    boolean hasGenericIndex = fmt.getInputFlag(IContainerFormat.Flags.FLAG_GENERIC_INDEX);
    assertTrue("should have global header", hasGenericIndex);
  }
  
  @Test
  public void testGetOutputFlag()
  {
    IContainerFormat fmt = IContainerFormat.make();
    
    int retval = -1;
    int flags = fmt.getOutputFlags();
    assertEquals("should be not set", flags, 0);
    retval = fmt.setOutputFormat("mov", null, null);
    assertTrue("should succeed", retval >= 0);
    boolean hasGlobalHeader = fmt.getOutputFlag(IContainerFormat.Flags.FLAG_GLOBALHEADER);
    assertTrue("should have global header", hasGlobalHeader);
  }
  
  @Test
  public void testSetInputFlag()
  {
    IContainerFormat fmt = IContainerFormat.make();
    int retval = -1;
    int flags = fmt.getInputFlags();
    assertEquals("should be not set", flags, 0);
    retval = fmt.setInputFormat("s16be");
    assertTrue("should succeed", retval >= 0);
    boolean hasGlobalHeader = fmt.getInputFlag(IContainerFormat.Flags.FLAG_GLOBALHEADER);
    assertTrue("should not have global header", !hasGlobalHeader);
    fmt.setInputFlag(IContainerFormat.Flags.FLAG_GLOBALHEADER, true);
    hasGlobalHeader = fmt.getInputFlag(IContainerFormat.Flags.FLAG_GLOBALHEADER);
    assertTrue("should have global header", hasGlobalHeader);

    fmt.setInputFlag(IContainerFormat.Flags.FLAG_GLOBALHEADER, false);
    hasGlobalHeader = fmt.getInputFlag(IContainerFormat.Flags.FLAG_GLOBALHEADER);
    assertTrue("should not have global header", !hasGlobalHeader);
  }
  
  @Test
  public void testSetOutputFlag()
  {
    IContainerFormat fmt = IContainerFormat.make();
    
    int retval = -1;
    int flags = fmt.getOutputFlags();
    assertEquals("should be not set", flags, 0);
    retval = fmt.setOutputFormat("mov", null, null);
    assertTrue("should succeed", retval >= 0);
    boolean hasGlobalHeader = fmt.getOutputFlag(IContainerFormat.Flags.FLAG_GLOBALHEADER);
    assertTrue("should have global header", hasGlobalHeader);
    
    fmt.setOutputFlag(IContainerFormat.Flags.FLAG_GLOBALHEADER, false);
    hasGlobalHeader = fmt.getOutputFlag(IContainerFormat.Flags.FLAG_GLOBALHEADER);
    assertTrue("should not have global header", !hasGlobalHeader);

    fmt.setOutputFlag(IContainerFormat.Flags.FLAG_GLOBALHEADER, true);
    hasGlobalHeader = fmt.getOutputFlag(IContainerFormat.Flags.FLAG_GLOBALHEADER);
    assertTrue("should have global header", hasGlobalHeader);
    

  }
  
  @Test
  public void testGetOutputCodecsSupported()
  {
    IContainerFormat fmt = IContainerFormat.make();
    
    int retval = -1;
    int flags = fmt.getOutputFlags();
    assertEquals("should be not set", flags, 0);
    retval = fmt.setOutputFormat("mov", null, null);
    assertTrue("should succeed", retval >= 0);
    
    List<ICodec.ID> codecs = fmt.getOutputCodecsSupported();
    assertNotNull(codecs);
//    for(ICodec.ID id : codecs)
//      System.out.println("Codec: "+id);

    assertTrue("should get at least one codec", codecs.size() > 1);
    assertTrue("should have MP3",
        codecs.contains(ICodec.ID.CODEC_ID_MP3));
    assertTrue("Should contain H263",
        codecs.contains(ICodec.ID.CODEC_ID_H263));
  }
  
  @Test
  public void testGetInputFormats()
  {
    Collection<IContainerFormat> installed =
      IContainerFormat.getInstalledInputFormats();
    assertTrue(installed.size() > 0);
    for(IContainerFormat fmt : installed)
    {
      assertNotNull(fmt);
      assertTrue(fmt.isInput());
      assertTrue(fmt.getInputFormatShortName().length() > 0);
    }
  }
  @Test
  public void testGetOutputFormats()
  {
    Collection<IContainerFormat> installed =
      IContainerFormat.getInstalledOutputFormats();
    assertTrue(installed.size() > 0);
    for(IContainerFormat fmt : installed)
    {
      assertNotNull(fmt);
      assertTrue(fmt.isOutput());
      assertTrue(fmt.getOutputFormatShortName().length() > 0);
    }
  }
  
  @Test
  public void testEstablishOutputCodecId()
  {
    JNIMemoryManager.getMgr().flush();
    IContainerFormat fmt = IContainerFormat.make();
    fmt.setOutputFormat("flv", null, null);
    assertEquals(ICodec.ID.CODEC_ID_FLV1,
        fmt.establishOutputCodecId(ICodec.Type.CODEC_TYPE_VIDEO));
    assertEquals(ICodec.ID.CODEC_ID_MP3,
        fmt.establishOutputCodecId(ICodec.Type.CODEC_TYPE_AUDIO));
    
    fmt.setOutputFormat("mp4", null, null);
    assertEquals(ICodec.ID.CODEC_ID_MPEG4,
        fmt.establishOutputCodecId(ICodec.Type.CODEC_TYPE_VIDEO));
    assertEquals(ICodec.ID.CODEC_ID_AAC,
        fmt.establishOutputCodecId(ICodec.Type.CODEC_TYPE_AUDIO));

    fmt.setOutputFormat("3gp", null, null);
    assertEquals(ICodec.ID.CODEC_ID_H263,
        fmt.establishOutputCodecId(ICodec.Type.CODEC_TYPE_VIDEO));
    assertEquals(ICodec.ID.CODEC_ID_AAC,
        fmt.establishOutputCodecId(ICodec.Type.CODEC_TYPE_AUDIO));

    fmt.delete();
    assertEquals(0, JNIMemoryManager.getMgr().getNumPinnedObjects());
  }
  
  @Test
  public void testEstablishOutputCodecIdFailOnMismatchedArgs()
  {
    JNIMemoryManager.getMgr().flush();
    IContainerFormat fmt = IContainerFormat.make();
    fmt.setOutputFormat("flv", null, null);
    try {
      fmt.establishOutputCodecId(ICodec.Type.CODEC_TYPE_VIDEO,
          ICodec.ID.CODEC_ID_MP3);
      fail("should not get here");
    } catch (IllegalArgumentException e) {}
    fmt.delete();
    assertEquals(0, JNIMemoryManager.getMgr().getNumPinnedObjects());
  }
  @Test
  public void testEstablishOutputCodecIdFailOnInputFormat()
  {
    JNIMemoryManager.getMgr().flush();
    IContainerFormat fmt = IContainerFormat.make();
    fmt.setInputFormat("flv");
    try {
      fmt.establishOutputCodecId(ICodec.Type.CODEC_TYPE_VIDEO);
      fail("should not get here");
    } catch (IllegalArgumentException e) {}
    fmt.delete();
    assertEquals(0, JNIMemoryManager.getMgr().getNumPinnedObjects());
  }


}
